qujax/gates.py

```diff
@@ -69,14 +96,90 @@ def Rz(param: float) -> jnp.ndarray:
     return jnp.cos(param_pi_2) * I - jnp.sin(param_pi_2) * Z * 1.j
 
 
+def CRx(param: float) -> jnp.ndarray:
+    return jnp.block([[I, _0],
+                      [_0, Rx(param)]]).reshape((2,) * 4)
+
+
+def CRy(param: float) -> jnp.ndarray:
+    return jnp.block([[I, _0],
+                      [_0, Ry(param)]]).reshape((2,) * 4)
+
+
+def CRz(param: float) -> jnp.ndarray:
+    return jnp.block([[I, _0],
+                      [_0, Rz(param)]]).reshape((2,) * 4)
+
+
 def U1(param: float) -> jnp.ndarray:
     return U3(0, 0, param)
 
 
-def U2(param1: float, param2: float) -> jnp.ndarray:
-    return U3(0.5, param1, param2)
+def U2(param0: float, param1: float) -> jnp.ndarray:
+    return U3(0.5, param0, param1)
+
+
+def U3(param0: float, param1: float, param2: float) -> jnp.ndarray:
+    return jnp.exp((param1 + param2) * jnp.pi * 1.j / 2) * Rz(param1) @ Ry(param0) @ Rz(param2)
 
 
-def U3(param1: float, param2: float, param3: float) -> jnp.ndarray:
-    return jnp.exp((param2 + param3) * jnp.pi * 1.j / 2) * Rz(param2) @ Ry(param1) @ Rz(param3)
+def CU1(param: float) -> jnp.ndarray:
+    return jnp.block([[I, _0],
+                      [_0, U1(param)]]).reshape((2,) * 4)
 
+
+def CU2(param0: float, param1: float) -> jnp.ndarray:
+    return jnp.block([[I, _0],
+                      [_0, U2(param0, param1)]]).reshape((2,) * 4)
+
+
+def CU3(param0: float, param1: float, param2: float) -> jnp.ndarray:
+    return jnp.block([[I, _0],
+                      [_0, U3(param0, param1, param2)]]).reshape((2,) * 4)
+
+
+def ISWAP(param: float) -> jnp.ndarray:
+    param_pi_2 = param * jnp.pi / 2
+    c = jnp.cos(param_pi_2)
+    i_s = 1.j * jnp.sin(param_pi_2)
+    return jnp.array([[1., 0., 0., 0.],
+                      [0., c, i_s, 0.],
+                      [0., i_s, c, 0.],
+                      [0., 0., 0., 1.]]).reshape((2,) * 4)
+
+
+def PhasedISWAP(param0: float, param1: float) -> jnp.ndarray:
+    param1_pi_2 = param1 * jnp.pi / 2
+    c = jnp.cos(param1_pi_2)
+    i_s_e = 1.j * jnp.sin(param1_pi_2) * jnp.exp(-2.j * jnp.pi * param0)
+    return jnp.array([[1., 0., 0., 0.],
+                      [0., c, i_s_e, 0.],
```

Comment:
this term should have a different sign in the exponential, no?

Reply:
Good spot! Now fixed